elastic/integrations
# [aws] [cloudwatch_metrics] Map aws.dimensions field as object

## Proposed commit message



Change the mapping type for the `aws.dimensions` field from `flattened` to `object`.

Currently, all `*_metrics` data streams but one use the `object` mapping. The `cloudwatch_metrics` data stream uses the `flattened` type instead.

We need to unify the mapping of `aws.dimensions` across all metrics-related data streams in the AWS integration.
 If all data streams use the exact mapping for `aws.dimensions`, users will be able to query and build a dashboard that correlates data across different data streams.

## Checklist


## Author's Checklist


- [ ] #11893 (to get `subobjects: false` support)
- [ ] Add a case for "How to test this PR locally" to prove the new mapping makes a difference for query and dashboard purposes.

## How to test this PR locally



- Start the local stack (use 8.17.0)
- Install the latest released integration (using AWS 2.36.2)
- Check the component template
- Post the test document
- Try to visualize the metric in Kibana
- Upgrade the integration to this PR version
- Delete the data stream and re-index the same test document
- Re-try to visualize the metric in Kibana using the new mapping

### Check the component template

With 2.36.2, the component template should be something similar to:

```text
// GET _component_template/metrics-aws.cloudwatch_metrics@package
{
  "aws": {
    "dynamic": true,
    "type": "object",
    "properties": {
      "dimensions_fingerprint": {
        "time_series_dimension": true,
        "type": "keyword"
      },
      "cloudwatch": {
        "properties": {
          "namespace": {
            "ignore_above": 1024,
            "type": "keyword"
          }
        }
      },
      "tags": {
        "type": "flattened"
      },
      "dimensions": {
        "type": "flattened"
      }
    }
  }
}
```

### Post the test document

Using the Dev Tools, post the following test document:

```text
// POST metrics-aws.cloudwatch_metrics-sdh5390/_doc
{
  "@timestamp": "2024-12-30T14:57:07+01:00",
    "aws": {
      "dimensions": {
        "name": "Maurizio Branca",
        "AutoScalingGroupName": "whatever"
        // "ImageId": "whatever",
        // "InstanceId": "whatever"
        //"InstanceType": "whatever"
      },
      "metric": {
        "cpu": 10
      }
  }
}
```

### Try to visualize the metric in Kibana

If I try to visualize the CPU metric in Kibana, the dimension subfields in `aws.dimensions.*` are not available:

![CleanShot 2024-12-30 at 16 53 42@2x](https://github.com/user-attachments/assets/170cf62c-167d-4b57-b4e1-4cb4f24ce7f9)

### Upgrade the integration to this PR version

Bump the integration version (for example, 2.37.0), and build and update the integration:

```sh
elastic-package build && elastic-package stack up -d -v --services package-registry
```

### Delete the data stream and re-index the same test document

```text
// DELETE _data_stream/metrics-aws.cloudwatch_metrics-sdh5390

// POST metrics-aws.cloudwatch_metrics-sdh5390/_doc
{
  "@timestamp": "2024-12-30T14:57:07+01:00",
    "aws": {
      "dimensions": {
        "name": "Maurizio Branca",
        "AutoScalingGroupName": "whatever"
        // "ImageId": "whatever",
        // "InstanceId": "whatever"
        //"InstanceType": "whatever"
      },
      "metric": {
        "cpu": 10
      }
  }
}
```

If we check the mapping, we should now see a dynamic template:

![CleanShot 2024-12-30 at 16 59 59@2x](https://github.com/user-attachments/assets/cc8bb392-3699-4c9b-9a0c-de037b72e879)

### Re-try to visualize the metric in Kibana using the new mapping

With the updated mapping, we can now use `aws.dimensions.*` fields in Kibana to break down values by a dimension:

![CleanShot 2024-12-30 at 17 01 32@2x](https://github.com/user-attachments/assets/a3b6c008-0bad-4d90-bc5a-50963c6d709f)


## Related issues


- Closes https://github.com/elastic/integrations/issues/11806
